sycl/plugins/unified_runtime/ur/adapters/cuda/memory.cpp

```diff
@@ -485,3 +482,90 @@ UR_APIEXPORT ur_result_t UR_APICALL urMemBufferPartition(
   *phMem = MemObj.release();
   return UR_RESULT_SUCCESS;
 }
+
+ur_result_t
+ur_buffer_::allocateMemObjOnDeviceIfNeeded(ur_device_handle_t hDevice) {
+  UR_ASSERT(hDevice, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+
+  ScopedDevice Active(hDevice);
+  ur_lock_guard LockGuard(MemoryAllocationMutex);
+
+  ur_result_t Result = UR_RESULT_SUCCESS;
+  CUdeviceptr &DevPtr = getNativePtr(hDevice);
+
+  // Allocation has already been made
+  if (DevPtr != ur_buffer_::native_type{0}) {
+    return UR_RESULT_SUCCESS;
+  }
+
+  if (MemAllocMode == ur_buffer_::AllocMode::AllocHostPtr) {
+    // Host allocation has already been made
+    Result = UR_CHECK_ERROR(cuMemHostGetDevicePointer(&DevPtr, HostPtr, 0));
+  } else if (MemAllocMode == ur_buffer_::AllocMode::UseHostPtr) {
+    Result = UR_CHECK_ERROR(
+        cuMemHostRegister(HostPtr, Size, CU_MEMHOSTREGISTER_DEVICEMAP));
+    Result = UR_CHECK_ERROR(cuMemHostGetDevicePointer(&DevPtr, HostPtr, 0));
+  } else {
+    Result = UR_CHECK_ERROR(cuMemAlloc(&DevPtr, Size));
+  }
+  return Result;
+}
+
+ur_result_t
+ur_image_::allocateMemObjOnDeviceIfNeeded(ur_device_handle_t hDevice) {
+  UR_ASSERT(hDevice, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  // TODO: Implement for Images
+  return UR_RESULT_SUCCESS;
+}
+
+// We should have already waited on the necessary events before calling this
+// function
+ur_result_t
+ur_buffer_::migrateMemoryToDeviceIfNeeded(ur_device_handle_t hDevice) {
+  UR_ASSERT(hDevice, UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+  UR_ASSERT(Ptrs[hDevice->getIndex()], UR_RESULT_ERROR_INVALID_NULL_HANDLE);
+
+  // Device allocation has already been initialized with most up to date
+  // data in buffer
+  if (HaveMigratedToDeviceSinceLastWrite[hDevice->getIndex()]) {
+    return UR_RESULT_SUCCESS;
+  }
+
+  ScopedDevice Active(hDevice);
+
+  ur_result_t Result = UR_RESULT_SUCCESS;
+
+  // If no kernels have written to the memobj then initialize the device
+  // allocation from host if it has not been initialized already
+  if (LastEventWritingToMemObj == nullptr) {
+    // Device allocation being initialized from host for the first time
+    if (HostPtr) {
+      Result = UR_CHECK_ERROR(
```

Comment:
Same re `UR_CHECK_ERROR` (and `if` based on `Result`).

Reply:
Yep nice thanks